packages/react/src/useActionForm.ts

```diff
@@ -0,0 +1,293 @@
+    "modelApiIdentifier" in action ? useAction(action, { select: options?.select } as any) : useGlobalAction(action);
+
+  const handleSubmissionError = useCallback(
+    (error: Error | FieldErrors<ActionFunc["variablesType"]>) => {
+      if ("executionErrors" in error) {
+        const executionErrors = (error as unknown as ErrorWrapper).executionErrors;
+        for (const executionError of executionErrors) {
+          if ("validationErrors" in executionError) {
+            for (const validationError of executionError.validationErrors) {
+              const errorKey =
+                "modelApiIdentifier" in action
+                  ? `${action.modelApiIdentifier}.${validationError.apiIdentifier}`
+                  : validationError.apiIdentifier;
+              formHook.setError(errorKey as any, {
+                message: validationError.message,
+              });
+            }
+          } else {
+            const codeToReplace = "code" in executionError ? `${executionError.code}: ` : "";
+            const message = executionError.message.replace(codeToReplace, "");
+
+            formHook.setError("root", { message });
+          }
+        }
+      }
+
+      options?.onError?.(error);
+    },
+    [action, formHook, options]
+  );
+
+  const submit = useCallback(
+    async (event?: React.BaseSyntheticEvent<object, any, any> | undefined) => {
+      let result: any;
+
+      formHook.clearErrors();
+
+      await handleSubmit(
+        async (data) => {
+          let variables: ActionFunc["variablesType"] = {
+            id: existingRecordId,
+            ...data,
+          };
+
+          if (options?.send) {
+            const unmasked = variables;
+            variables = {};
+            for (const key of options.send) {
+              set(variables, key, get(unmasked, key));
+            }
+          }
+
+          options?.onSubmit?.();
+
+          result = await runAction(variables);
+          if (!result.error) {
+            options?.onSuccess?.(result.data);
+          }
+        },
+        (errors) => {
+          handleSubmissionError(errors);
+        }
+      )(event);
+
+      if (result?.error) {
+        handleSubmissionError(result.error);
+      }
+
+      return result;
+    },
+    [handleSubmit, formHook, handleSubmissionError, existingRecordId, options, runAction]
+  );
+
+  const proxiedFormState = new Proxy(formState, {
+    get: (target, prop) => {
+      if (prop === "isSubmitting") {
+        return target.isSubmitting || actionResult.fetching;
+      } else if (prop === "isSubmitSuccessful") {
+        return target.isSubmitSuccessful && !actionResult.fetching && !actionResult.error;
+      } else if (prop === "isLoading") {
+        return target.isLoading || findResult.fetching;
+      } else {
+        return (target as any)[prop];
+      }
+    },
+  });
+
+  return {
+    ...formHook,
+    formState: proxiedFormState,
+    error: findResult.error || actionResult.error,
```

Comment:
Do we need to return this if we're going to put errors at the root? Kind of annoying to make the user check two places?

Reply:
I think since we only show the message on the formState (and also I haven't put the `findResult.error` there) it would be useful to have the full errors as well